### DISCUSSION TITLE
Polymorphic chargeable association on Pay::Charge

### DISCUSSION THREAD

**Discussion by mguidetti** [pay-rails/pay#609] [Ideas]
In my application I have two models that can be related to one-off charges, Invoice and FormResponse. Currently, I associate these other records by adding has_one relationships to Pay::Charge with an extension:

```
included do
  has_one :invoice
  has_one :form_response
end
```

with an accompanying belongs_to on the Invoice and FormResponse models.

This works, but what would you think about adding a polymorphic "Chargable" association on Pay::Charge? This would allow a built-in way to associate a Pay::Charge record to any other record in an application.

I'd imagine usage being something like:

```
@user.payment_processor.charge(15_00, chargeable: SomeChargeableRecordInstance)
```

If this seems like a good idea, I'd be happy to start PR.

**Comment by excid3** [comment]
There's definitely something that could be good here, although I don't think adding it to the Pay::Charge makes sense since it limits to a single association.

ActiveStorage is a good example of how it should ideally work for maximum use cases:

`Blob` - the actual uploaded file
- `has_many :attachments`

`Attachment`
- `belongs_to :blob`
- `belongs_to :record, polymorphic: true, touch: true`

You could apply this approach to both `Pay::Subscriptions` and `Pay::Charges` too. Maybe that's a model that should be added to Pay, I'm not sure. `Pay::Attachment`?

**Comment by mguidetti** [comment]
Right that's a good idea to allow multiple associations. I'll start that way and see what I can come up with. 👍
